### DISCUSSION TITLE
AsyncCloseException: Connection closed

### DISCUSSION THREAD

**Discussion by CarlSayres** [activej/activej#254] [Q&A]
Hi,
I am using ActiveJ RPC with multiple RPC servers behind a network load balancer. This works well, however the NLB closes connections when there are periods of inactivity. When a client tries to make an RPC call, we get the following exception:
```
io.activej.async.exception.AsyncCloseException: Connection closed
  java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357)
  java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1908)
```
I can mitigate this through retrying. But I'm wondering if ActiveJ has a mechanism to automatically reconnect a closed connection?

Thanks,
Carl

**Comment by eduard-vasinskyi** [comment]
Hi, @CarlSayres

ActiveJ RPC does not create a connection for each request, but maintains a persistent connection to the server (multiple servers). In case a connection is closed, RpcClient automatically tries to reconnect to the server. However, since TCP is a passive protocol, there is no way to tell that the connection is closed until you try to read/write to the socket. 

In order to keep RPC connection from going idle, you need to create RpcClient with keep-alive option: `RpcClient#withKeepAlive(Duration)`. Here you can specify an interval for Ping-Pong messages to be exchanged between RPC client and server. This message exchange will keep the connection from going idle. You need to set the interval lower than the load balancer’s idle timeout.
